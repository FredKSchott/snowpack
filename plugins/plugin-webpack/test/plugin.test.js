const path = require('path');
const fs = require('fs-extra');

const plugin = require('../plugin');
const readFilesSync = require('./readFilesSync');

const STUBS_DIR = path.join(__dirname, 'stubs/minimal/');
const IGNORED_STUBS_DIR = path.join(__dirname, 'stubs/minimal_ignore/');

const MULTIPLE_DIR = path.join(__dirname, 'stubs/multiple-entrypoints/');
const IGNORED_MULTIPLE_DIR = path.join(__dirname, 'stubs/multiple-entrypoints_ignore/');

describe('@snowpack/plugin-webpack', () => {
  // Copy over the stub folder to an git-ignored path and mock console.log
  beforeEach(() => {
    if (fs.existsSync(IGNORED_STUBS_DIR)) fs.removeSync(IGNORED_STUBS_DIR);
    fs.copySync(STUBS_DIR, IGNORED_STUBS_DIR);

    if (fs.existsSync(IGNORED_MULTIPLE_DIR)) fs.removeSync(IGNORED_MULTIPLE_DIR);
    fs.copySync(MULTIPLE_DIR, IGNORED_MULTIPLE_DIR);
  });

  afterAll(() => {
    if (fs.existsSync(IGNORED_STUBS_DIR)) fs.removeSync(IGNORED_STUBS_DIR);
    if (fs.existsSync(IGNORED_MULTIPLE_DIR)) fs.removeSync(IGNORED_MULTIPLE_DIR);
  });

  it('minimal - no options', async () => {
    const pluginInstance = plugin({
      buildOptions: {},
    });

    await pluginInstance.optimize({
      buildDirectory: IGNORED_STUBS_DIR,
    });

    expect(readFilesSync(IGNORED_STUBS_DIR)).toMatchSnapshot('files');
  });

  it('minimal - all options', async () => {
    const pluginInstance = plugin(
      {
        buildOptions: {},
      },
      {
        sourceMap: true,
        outputPattern: {
          js: '[name]-[id].js',
        },
        extendConfig: (config) => config,
        manifest: true,
        htmlMinifierOptions: true,
      },
    );

    await pluginInstance.optimize({
      buildDirectory: IGNORED_STUBS_DIR,
    });

    expect(readFilesSync(IGNORED_STUBS_DIR)).toMatchSnapshot('files');
  });

  it('multiple entrypoints w/ same filename', async () => {
    const pluginInstance = plugin({
      buildOptions: {},
    });

    await pluginInstance.optimize({
      buildDirectory: IGNORED_MULTIPLE_DIR,
    });

    const rootHtml = fs.readFileSync(path.join(IGNORED_MULTIPLE_DIR, 'index.html'), { encoding: 'utf8' });
    const adminHtml = fs.readFileSync(path.join(IGNORED_MULTIPLE_DIR, 'admin/index.html'), { encoding: 'utf8' });

    const rootScripts = rootHtml.match(/<script src="([^>]+)">/g);
    const adminScripts = adminHtml.match(/<script src="([^>]+)">/g);

    expect(rootScripts).not.toEqual(expect.arrayContaining(adminScripts));

    expect(readFilesSync(IGNORED_MULTIPLE_DIR)).toMatchSnapshot('files');
    expect(readFilesSync(path.join(IGNORED_MULTIPLE_DIR, 'admin'))).toMatchSnapshot('files');
  });
});
