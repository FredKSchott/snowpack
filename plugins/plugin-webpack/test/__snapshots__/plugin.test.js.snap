// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@snowpack/plugin-webpack minimal - all options: files 1`] = `
Array [
  Object {
    "content": "{
  \\"index.js\\": \\"/index-826.js\\",
  \\"webpack-runtime.js\\": \\"/webpack-runtime-658.js\\",
  \\"styles.css\\": \\"/css/styles.XXXXXXXXXXXXXXXXXXXX.css\\",
  \\"assets/backpack.svg\\": \\"/assets/backpack.XXXXXXXXXXXXXXXXXXXX.svg\\",
  \\"index-826.js.map\\": \\"/index-826.js.map\\",
  \\"webpack-runtime-658.js.map\\": \\"/webpack-runtime-658.js.map\\",
  \\"styles.css.map\\": \\"/css/styles.XXXXXXXXXXXXXXXXXXXX.css.map\\"
}",
    "filename": "asset-manifest.json",
  },
  Object {
    "content": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 20 20\\"><path d=\\"M14.707 12.293a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 111.414-1.414L9 14.586V3a1 1 0 012 0v11.586l2.293-2.293a1 1 0 011.414 0z\\"/></svg>
",
    "filename": "backpack.svg",
  },
  Object {
    "content": "(self.webpackChunkplugin_webpack_test_minimal=self.webpackChunkplugin_webpack_test_minimal||[]).push([[826],{99:(s,e,t)=>{\\"use strict\\";t(933),console.log(\\"test\\")},933:(s,e,t)=>{\\"use strict\\";s.exports=t.p+\\"assets/backpack.XXXXXXXXXXXXXXXXXXXX.svg\\"}},s=>{\\"use strict\\";s.O(0,[532],(()=>(99,s(s.s=99)))),s.O()}]);
//# sourceMappingURL=index-826.js.map",
    "filename": "index-826.js",
  },
  Object {
    "content": "{\\"version\\":3,\\"sources\\":[\\"webpack://plugin-webpack-test-minimal/./index.js\\"],\\"names\\":[\\"console\\",\\"log\\"],\\"mappings\\":\\"8IAEAA,QAAQC,IAAI,S\\",\\"file\\":\\"index-826.js\\",\\"sourcesContent\\":[\\"import './styles.css';\\\\nimport './backpack.svg';\\\\nconsole.log('test');\\\\n\\"],\\"sourceRoot\\":\\"\\"}",
    "filename": "index-826.js.map",
  },
  Object {
    "content": "<!DOCTYPE html><html lang=\\"en\\"><head><meta charset=\\"UTF-8\\"><title>Minimal test</title><link rel=\\"stylesheet\\" href=\\"/css/styles.XXXXXXXXXXXXXXXXXXXX.css\\"></head><body><script src=\\"/webpack-runtime-658.js\\" async=\\"\\" defer=\\"\\"></script><script src=\\"/index-826.js\\" async=\\"\\" defer=\\"\\"></script></body></html>",
    "filename": "index.html",
  },
  Object {
    "content": "import './styles.css';
import './backpack.svg';
console.log('test');
",
    "filename": "index.js",
  },
  Object {
    "content": "{
  \\"name\\": \\"plugin-webpack-test-minimal\\",
  \\"main\\": \\"src/index.js\\"
}
",
    "filename": "package.json",
  },
  Object {
    "content": "body {
  background-color: skyblue;
}
",
    "filename": "styles.css",
  },
  Object {
    "content": "(()=>{\\"use strict\\";var r,e={},t={};function n(r){var a=t[r];if(void 0!==a)return a.exports;var l=t[r]={exports:{}};return e[r](l,l.exports,n),l.exports}n.m=e,r=[],n.O=(e,t,a,l)=>{if(!t){var o=1/0;for(s=0;s<r.length;s++){for(var[t,a,l]=r[s],p=!0,i=0;i<t.length;i++)(!1&l||o>=l)&&Object.keys(n.O).every((r=>n.O[r](t[i])))?t.splice(i--,1):(p=!1,l<o&&(o=l));p&&(r.splice(s--,1),e=a())}return e}l=l||0;for(var s=r.length;s>0&&r[s-1][2]>l;s--)r[s]=r[s-1];r[s]=[t,a,l]},n.o=(r,e)=>Object.prototype.hasOwnProperty.call(r,e),n.p=\\"/\\",(()=>{var r={658:0,532:0};n.O.j=e=>0===r[e];var e=(e,t)=>{var a,l,[o,p,i]=t,s=0;for(a in p)n.o(p,a)&&(n.m[a]=p[a]);for(i&&i(n),e&&e(t);s<o.length;s++)l=o[s],n.o(r,l)&&r[l]&&r[l][0](),r[o[s]]=0;n.O()},t=self.webpackChunkplugin_webpack_test_minimal=self.webpackChunkplugin_webpack_test_minimal||[];t.forEach(e.bind(null,0)),t.push=e.bind(null,t.push.bind(t))})(),n.O()})();
//# sourceMappingURL=webpack-runtime-658.js.map",
    "filename": "webpack-runtime-658.js",
  },
  Object {
    "content": "{\\"version\\":3,\\"sources\\":[\\"webpack://plugin-webpack-test-minimal/webpack/runtime/chunk loaded\\",\\"webpack://plugin-webpack-test-minimal/webpack/bootstrap\\",\\"webpack://plugin-webpack-test-minimal/webpack/runtime/hasOwnProperty shorthand\\",\\"webpack://plugin-webpack-test-minimal/webpack/runtime/publicPath\\",\\"webpack://plugin-webpack-test-minimal/webpack/runtime/jsonp chunk loading\\"],\\"names\\":[\\"deferred\\",\\"__webpack_module_cache__\\",\\"__webpack_require__\\",\\"moduleId\\",\\"cachedModule\\",\\"undefined\\",\\"exports\\",\\"module\\",\\"__webpack_modules__\\",\\"m\\",\\"O\\",\\"result\\",\\"chunkIds\\",\\"fn\\",\\"priority\\",\\"notFulfilled\\",\\"Infinity\\",\\"i\\",\\"length\\",\\"fulfilled\\",\\"j\\",\\"Object\\",\\"keys\\",\\"every\\",\\"key\\",\\"splice\\",\\"o\\",\\"obj\\",\\"prop\\",\\"prototype\\",\\"hasOwnProperty\\",\\"call\\",\\"p\\",\\"installedChunks\\",\\"658\\",\\"532\\",\\"chunkId\\",\\"webpackJsonpCallback\\",\\"parentChunkLoadingFunction\\",\\"data\\",\\"moreModules\\",\\"runtime\\",\\"chunkLoadingGlobal\\",\\"self\\",\\"forEach\\",\\"bind\\",\\"push\\"],\\"mappings\\":\\"uBAAIA,E,KCCAC,EAA2B,GAG/B,SAASC,EAAoBC,GAE5B,IAAIC,EAAeH,EAAyBE,GAC5C,QAAqBE,IAAjBD,EACH,OAAOA,EAAaE,QAGrB,IAAIC,EAASN,EAAyBE,GAAY,CAGjDG,QAAS,IAOV,OAHAE,EAAoBL,GAAUI,EAAQA,EAAOD,QAASJ,GAG/CK,EAAOD,QAIfJ,EAAoBO,EAAID,EDzBpBR,EAAW,GACfE,EAAoBQ,EAAI,CAACC,EAAQC,EAAUC,EAAIC,KAC9C,IAAGF,EAAH,CAMA,IAAIG,EAAeC,IACnB,IAASC,EAAI,EAAGA,EAAIjB,EAASkB,OAAQD,IAAK,CAGzC,IAFA,IAAKL,EAAUC,EAAIC,GAAYd,EAASiB,GACpCE,GAAY,EACPC,EAAI,EAAGA,EAAIR,EAASM,OAAQE,MACpB,EAAXN,GAAsBC,GAAgBD,IAAaO,OAAOC,KAAKpB,EAAoBQ,GAAGa,OAAOC,GAAStB,EAAoBQ,EAAEc,GAAKZ,EAASQ,MAC9IR,EAASa,OAAOL,IAAK,IAErBD,GAAY,EACTL,EAAWC,IAAcA,EAAeD,IAG1CK,IACFnB,EAASyB,OAAOR,IAAK,GACrBN,EAASE,KAGX,OAAOF,EAtBNG,EAAWA,GAAY,EACvB,IAAI,IAAIG,EAAIjB,EAASkB,OAAQD,EAAI,GAAKjB,EAASiB,EAAI,GAAG,GAAKH,EAAUG,IAAKjB,EAASiB,GAAKjB,EAASiB,EAAI,GACrGjB,EAASiB,GAAK,CAACL,EAAUC,EAAIC,IEL/BZ,EAAoBwB,EAAI,CAACC,EAAKC,IAAUP,OAAOQ,UAAUC,eAAeC,KAAKJ,EAAKC,GCAlF1B,EAAoB8B,EAAI,I,MCKxB,IAAIC,EAAkB,CACrBC,IAAK,EACLC,IAAK,GAaNjC,EAAoBQ,EAAEU,EAAKgB,GAA0C,IAA7BH,EAAgBG,GAGxD,IAAIC,EAAuB,CAACC,EAA4BC,KACvD,IAGIpC,EAAUiC,GAHTxB,EAAU4B,EAAaC,GAAWF,EAGhBtB,EAAI,EAC3B,IAAId,KAAYqC,EACZtC,EAAoBwB,EAAEc,EAAarC,KACrCD,EAAoBO,EAAEN,GAAYqC,EAAYrC,IAKhD,IAFGsC,GAASA,EAAQvC,GACjBoC,GAA4BA,EAA2BC,GACrDtB,EAAIL,EAASM,OAAQD,IACzBmB,EAAUxB,EAASK,GAChBf,EAAoBwB,EAAEO,EAAiBG,IAAYH,EAAgBG,IACrEH,EAAgBG,GAAS,KAE1BH,EAAgBrB,EAASK,IAAM,EAEhCf,EAAoBQ,KAGjBgC,EAAqBC,KAA8C,wCAAIA,KAA8C,yCAAK,GAC9HD,EAAmBE,QAAQP,EAAqBQ,KAAK,KAAM,IAC3DH,EAAmBI,KAAOT,EAAqBQ,KAAK,KAAMH,EAAmBI,KAAKD,KAAKH,K\\",\\"file\\":\\"webpack-runtime-658.js\\",\\"sourcesContent\\":[\\"var deferred = [];\\\\n__webpack_require__.O = (result, chunkIds, fn, priority) => {\\\\n\\\\tif(chunkIds) {\\\\n\\\\t\\\\tpriority = priority || 0;\\\\n\\\\t\\\\tfor(var i = deferred.length; i > 0 && deferred[i - 1][2] > priority; i--) deferred[i] = deferred[i - 1];\\\\n\\\\t\\\\tdeferred[i] = [chunkIds, fn, priority];\\\\n\\\\t\\\\treturn;\\\\n\\\\t}\\\\n\\\\tvar notFulfilled = Infinity;\\\\n\\\\tfor (var i = 0; i < deferred.length; i++) {\\\\n\\\\t\\\\tvar [chunkIds, fn, priority] = deferred[i];\\\\n\\\\t\\\\tvar fulfilled = true;\\\\n\\\\t\\\\tfor (var j = 0; j < chunkIds.length; j++) {\\\\n\\\\t\\\\t\\\\tif ((priority & 1 === 0 || notFulfilled >= priority) && Object.keys(__webpack_require__.O).every((key) => (__webpack_require__.O[key](chunkIds[j])))) {\\\\n\\\\t\\\\t\\\\t\\\\tchunkIds.splice(j--, 1);\\\\n\\\\t\\\\t\\\\t} else {\\\\n\\\\t\\\\t\\\\t\\\\tfulfilled = false;\\\\n\\\\t\\\\t\\\\t\\\\tif(priority < notFulfilled) notFulfilled = priority;\\\\n\\\\t\\\\t\\\\t}\\\\n\\\\t\\\\t}\\\\n\\\\t\\\\tif(fulfilled) {\\\\n\\\\t\\\\t\\\\tdeferred.splice(i--, 1)\\\\n\\\\t\\\\t\\\\tresult = fn();\\\\n\\\\t\\\\t}\\\\n\\\\t}\\\\n\\\\treturn result;\\\\n};\\",\\"// The module cache\\\\nvar __webpack_module_cache__ = {};\\\\n\\\\n// The require function\\\\nfunction __webpack_require__(moduleId) {\\\\n\\\\t// Check if module is in cache\\\\n\\\\tvar cachedModule = __webpack_module_cache__[moduleId];\\\\n\\\\tif (cachedModule !== undefined) {\\\\n\\\\t\\\\treturn cachedModule.exports;\\\\n\\\\t}\\\\n\\\\t// Create a new module (and put it into the cache)\\\\n\\\\tvar module = __webpack_module_cache__[moduleId] = {\\\\n\\\\t\\\\t// no module.id needed\\\\n\\\\t\\\\t// no module.loaded needed\\\\n\\\\t\\\\texports: {}\\\\n\\\\t};\\\\n\\\\n\\\\t// Execute the module function\\\\n\\\\t__webpack_modules__[moduleId](module, module.exports, __webpack_require__);\\\\n\\\\n\\\\t// Return the exports of the module\\\\n\\\\treturn module.exports;\\\\n}\\\\n\\\\n// expose the modules object (__webpack_modules__)\\\\n__webpack_require__.m = __webpack_modules__;\\\\n\\\\n\\",\\"__webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))\\",\\"__webpack_require__.p = \\\\\\"/\\\\\\";\\",\\"// no baseURI\\\\n\\\\n// object to store loaded and loading chunks\\\\n// undefined = chunk not loaded, null = chunk preloaded/prefetched\\\\n// [resolve, reject, Promise] = chunk loading, 0 = chunk loaded\\\\nvar installedChunks = {\\\\n\\\\t658: 0,\\\\n\\\\t532: 0\\\\n};\\\\n\\\\n// no chunk on demand loading\\\\n\\\\n// no prefetching\\\\n\\\\n// no preloaded\\\\n\\\\n// no HMR\\\\n\\\\n// no HMR manifest\\\\n\\\\n__webpack_require__.O.j = (chunkId) => (installedChunks[chunkId] === 0);\\\\n\\\\n// install a JSONP callback for chunk loading\\\\nvar webpackJsonpCallback = (parentChunkLoadingFunction, data) => {\\\\n\\\\tvar [chunkIds, moreModules, runtime] = data;\\\\n\\\\t// add \\\\\\"moreModules\\\\\\" to the modules object,\\\\n\\\\t// then flag all \\\\\\"chunkIds\\\\\\" as loaded and fire callback\\\\n\\\\tvar moduleId, chunkId, i = 0;\\\\n\\\\tfor(moduleId in moreModules) {\\\\n\\\\t\\\\tif(__webpack_require__.o(moreModules, moduleId)) {\\\\n\\\\t\\\\t\\\\t__webpack_require__.m[moduleId] = moreModules[moduleId];\\\\n\\\\t\\\\t}\\\\n\\\\t}\\\\n\\\\tif(runtime) runtime(__webpack_require__);\\\\n\\\\tif(parentChunkLoadingFunction) parentChunkLoadingFunction(data);\\\\n\\\\tfor(;i < chunkIds.length; i++) {\\\\n\\\\t\\\\tchunkId = chunkIds[i];\\\\n\\\\t\\\\tif(__webpack_require__.o(installedChunks, chunkId) && installedChunks[chunkId]) {\\\\n\\\\t\\\\t\\\\tinstalledChunks[chunkId][0]();\\\\n\\\\t\\\\t}\\\\n\\\\t\\\\tinstalledChunks[chunkIds[i]] = 0;\\\\n\\\\t}\\\\n\\\\t__webpack_require__.O();\\\\n}\\\\n\\\\nvar chunkLoadingGlobal = self[\\\\\\"webpackChunkplugin_webpack_test_minimal\\\\\\"] = self[\\\\\\"webpackChunkplugin_webpack_test_minimal\\\\\\"] || [];\\\\nchunkLoadingGlobal.forEach(webpackJsonpCallback.bind(null, 0));\\\\nchunkLoadingGlobal.push = webpackJsonpCallback.bind(null, chunkLoadingGlobal.push.bind(chunkLoadingGlobal));\\"],\\"sourceRoot\\":\\"\\"}",
    "filename": "webpack-runtime-658.js.map",
  },
]
`;

exports[`@snowpack/plugin-webpack minimal - no options: files 1`] = `
Array [
  Object {
    "content": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 20 20\\"><path d=\\"M14.707 12.293a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 111.414-1.414L9 14.586V3a1 1 0 012 0v11.586l2.293-2.293a1 1 0 011.414 0z\\"/></svg>
",
    "filename": "backpack.svg",
  },
  Object {
    "content": "<!DOCTYPE html><html lang=\\"en\\"><head><meta charset=\\"UTF-8\\"><title>Minimal test</title><link rel=\\"stylesheet\\" href=\\"/css/styles.XXXXXXXXXXXXXXXXXXXX.css\\"></head><body><script src=\\"/js/webpack-runtime.XXXXXXXXXXXXXXXXXXXX.js\\" async=\\"\\" defer=\\"\\"></script><script src=\\"/js/index.XXXXXXXXXXXXXXXXXXXX.js\\" async=\\"\\" defer=\\"\\"></script></body></html>",
    "filename": "index.html",
  },
  Object {
    "content": "import './styles.css';
import './backpack.svg';
console.log('test');
",
    "filename": "index.js",
  },
  Object {
    "content": "{
  \\"name\\": \\"plugin-webpack-test-minimal\\",
  \\"main\\": \\"src/index.js\\"
}
",
    "filename": "package.json",
  },
  Object {
    "content": "body {
  background-color: skyblue;
}
",
    "filename": "styles.css",
  },
]
`;

exports[`@snowpack/plugin-webpack multiple entrypoints w/ same filename: files 1`] = `
Array [
  Object {
    "content": "<!DOCTYPE html><html lang=\\"en\\"><head><meta charset=\\"UTF-8\\"><title>Multiple entrypoints test - root</title></head><body><script src=\\"/js/webpack-runtime.XXXXXXXXXXXXXXXXXXXX.js\\" async=\\"\\" defer=\\"\\"></script><script src=\\"/js/index.XXXXXXXXXXXXXXXXXXXX.js\\" async=\\"\\" defer=\\"\\"></script></body></html>",
    "filename": "index.html",
  },
  Object {
    "content": "console.log('root test');
",
    "filename": "index.js",
  },
  Object {
    "content": "{
  \\"name\\": \\"plugin-webpack-test-minimal\\",
  \\"main\\": \\"src/index.js\\"
}
",
    "filename": "package.json",
  },
]
`;

exports[`@snowpack/plugin-webpack multiple entrypoints w/ same filename: files 2`] = `
Array [
  Object {
    "content": "<!DOCTYPE html><html lang=\\"en\\"><head><meta charset=\\"UTF-8\\"><title>Multiple entrypoints test - /admin</title></head><body><script src=\\"/js/webpack-runtime.XXXXXXXXXXXXXXXXXXXX.js\\" async=\\"\\" defer=\\"\\"></script><script src=\\"/js/admin/index.XXXXXXXXXXXXXXXXXXXX.js\\" async=\\"\\" defer=\\"\\"></script></body></html>",
    "filename": "index.html",
  },
  Object {
    "content": "console.log('admin test');
",
    "filename": "index.js",
  },
]
`;
